ledger/simulation/resources.go

```diff
@@ -0,0 +1,391 @@
+	if ep.GetApplicationAddress(aid) == addr {
+		// The app local of an app and its own account is always available, so don't bother recording it.
+		return true
+	}
+	// nil map lookup is ok
+	_, ok := a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AppLocals == nil {
+		a.AppLocals = make(map[ledgercore.AccountApp]struct{})
+	}
+	a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}] = struct{}{}
+	return true
+}
+
+func (a *GroupResourceAssignment) boxIOBudget(bytesPerBoxRef uint64) uint64 {
+	return uint64(a.startingBoxes+a.Resources.maxPossibleBoxes()) * bytesPerBoxRef
+}
+
+type resourcePolicy struct {
+	assignment                  GroupResourceAssignment
+	ep                          *logic.EvalParams
+	initialBoxSurplusReadBudget *uint64
+	unnamedBoxUsedReadBudget    uint64
+
+	txnRootIndex   int
+	programVersion uint64
+	globalSharing  bool
+}
+
+func newResourcePolicy(ep *logic.EvalParams, groupResult *TxnGroupResult) *resourcePolicy {
+	policy := resourcePolicy{
+		assignment: makeGroupResourceAssignment(ep.TxnGroup, ep.Proto),
+		ep:         ep,
+	}
+	groupResult.UnnamedResources = &policy.assignment
+	for i := range groupResult.Txns {
+		groupResult.Txns[i].UnnamedResources = &policy.assignment.localTxnResources[i]
+	}
+	return &policy
+}
+
+func (p *resourcePolicy) AvailableAccount(addr basics.Address) bool {
+	if p.assignment.hasAccount(addr, p.ep, p.programVersion, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addAccount(addr, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AvailableAsset(aid basics.AssetIndex) bool {
+	if p.assignment.hasAsset(aid, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addAsset(aid, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AvailableApp(aid basics.AppIndex) bool {
+	if p.assignment.hasApp(aid, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addApp(aid, p.ep, p.programVersion, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AllowsHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// holdings are only checked if globalSharing is true
+	if p.assignment.hasHolding(addr, aid) {
+		return true
+	}
+	return p.assignment.addHolding(addr, aid)
+}
+
+func (p *resourcePolicy) AllowsLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// locals are only checked if globalSharing is true
+	if p.assignment.hasLocal(addr, aid, p.ep) {
+		return true
+	}
+	return p.assignment.addLocal(addr, aid)
+}
+
+func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string) bool {
+	if p.assignment.hasBox(app, name) {
+		return true
+	}
+	box, ok, err := p.ep.Ledger.GetBox(app, name)
+	if err != nil {
+		panic(err)
```

Comment:
question: is it possible to eat up the `err` here by returning just false?

or `err` implies something actually bad, so there is an `ok` return from `GetBox`?

Reply:
If `err` has a value, that means something really bad happened, like a DB error. If the box does not exist, we expect `ok` to be false and `err` to be nil.